### ISSUE TITLE
Replace `sprintf` with safer `snprintf` equivalents is DSP code

### ISSUE THREAD

**Issue by mborgerson** [xqemu/xqemu#174]
https://github.com/xqemu/xqemu/blob/01be8ef6b8f22897d614cb1b0876adcb572143cb/hw/xbox/dsp/dsp_cpu.c#L588

When DSP instruction tracing is enabled these usages of `snprintf` will cause buffer overflows.

**Comment by mborgerson** [comment]
*Apparently I forgot to reply to this comment...replying now as I check the backlog for low-hanging issues.*

Yes, adjusting the size of the buffers here would work around this issue, but I believe the bigger problem here is the unchecked use of `sprintf` functions--which led to a buffer overflow, in concert with a silly padding issue.

I'll explain further here what is happening and why I highlighted this as an example for why we should be using the safer alternative `snprintf`. The buffers in question are defined as:

```c
char disasm_str_instr[128];
char disasm_str_instr2[128];
```

Followed later by the section I previously highlighted:
```c
const int len = sizeof(dsp->disasm_str_instr);
...
offset = sprintf(dsp->disasm_str_instr2, "p:%04x  %06x         (%02d cyc)  %-*s\n", dsp->disasm_prev_inst_pc, dsp->disasm_cur_inst, dsp->instr_cycle, len, dsp->disasm_str_instr); 
```

The final `printf` fomatting token `%-*s` will print the string `dsp->disasm_str_instr`, left-aligned and padded to no fewer than `len` characters. `len` is always 128 == `sizeof(disasm_str_instr)`, resulting in an unavoidable overflow of the `disasm_str_instr2` buffer, which is also just 128 bytes.

It is true that if `disasm_str_instr2` were at least `sizeof(disasm_str_instr)` + whatever size we need to print the surrounding text (27 bytes or so), then this overflow wouldn't have happened. But, had we been using the safer alternative `snprintf`, this buffer overflow would not even be possible... though the code would have been just as silly to do this padding. There are other instances in the DSP code where it unscrupulously prints into a buffer without bounds checking, which I'd like to avoid as well.

So my suggestion here is the following:
* Remove the end of line padding in this instance of the highlighted format string.
* Replace instances of `sprintf` with the safer alternative `snprintf` as best practice to avoid future buffer overflows in the event such a stupid error is made again (which next time might be more difficult to discover).

**Comment by JayFoxRox** [comment]
> Yes, adjusting the size of the buffers here would work around this issue

I disagree. It's a fix, not a workaround.

> But, had we been using the safer alternative snprintf, this buffer overflow would not even be possible... 

It would avoid the overflow, but still cause errornous output (cut-off). So you are trading a bug for another bug.
So I consider  `snprintf` to be the workaround.

The fix would be to avoid the overflow by always having enough room for results (by providing sane input and acceptable buffer sizes). `snprintf` is just an additional protection for things we can't foresee at compile-time.

So my suggestion is:

- Check if upstream has already fixed his
- If not:
    - Fix things compiler warns about by bumping buffer sizes like `[16]` &rarr; `[32]`; particularly the ~40 warnings like this in dsp_dis.inl:
        ```
        hw/xbox/dsp/dsp_dis.inl:1058:9: note: ‘sprintf’ output between 3 and 18 bytes into a destination of size 16
             sprintf(srcname, "x:%s", addr_name);
        ```
    - Fix the mentioned disassembly alignment bug; removing padding is fine, as this is left-aligned anyway.
    - Additionally, `sprintf` &rarr; `snprintf`, and during development somehow report cases / review where trimming still occurred (I assume there's many instances).
    - Where trimming occurred, bump buffer sizes and / or fix logic.

**Comment by mborgerson** [comment]
> I disagree. It's a fix, not a workaround.

Yep, increasing the buffer sizes *will* make it so they aren't overflowed.

> It would avoid the overflow, but still cause erroneous output (cut-off). So you are trading a bug for another bug. So I consider snprintf to be the workaround.

It would cut off the padding, and the output in this case would be incorrect, but more importantly it would prevent the overflow of the buffer. It would be incorrect, but it wouldn't be catastrophically incorrect.

> Check if upstream has already fixed his [...] Also notify upstream of our improvements.

Good suggestion! We should be better generally about syncing with upstream for the DSP.

> Fix compiler warnings, by bumping buffer sizes like [16] → [32]; particularly the ~40 warnings like this in dsp_dis.inl: 
> Fix the mentioned disassembly alignment bug; removing padding is fine, as this is left-aligned anyway.
> Additionally, sprintf → snprintf, and during development somehow report cases / review where trimming still occurred (I assume there's many instances).
> Where trimming occurred, bump buffer sizes and / or fix logic.

It seems like we are in agreement on all of these things.
